2025-06-30: status update (#317)
nazar-pc/abundance: website/main/content/blog/2025-06-30-sector-expiration.md

need to expire in a way that does not require farmers to re-plot while archiving the most recent
history. Fortunately, we can leverage the current expiration mechanism of the current Subspace
protocol, and build a layer on top of it to adapt it to the sharded version while maintaining the
original guarantees in terms of plot expiration, sector re-plotting, and history archiving.

<!--more-->

## Sharded Subspace Plot Expiration Protocol

### Overview

Before I start with the description of the protocol, let me remind you what are the core properties
that we wanted the protocol to have:

1.  **Unique Plot IDs:** `plot_id` is unique and stable. This is key for the protocol, as this ID
    will be used to allocate space into shards.
2.  **No Bias through `history_size`:** Farmers can't game shard allocation by choosing among any
    `history_size`, or by creating sectors with the same ID committed to different history size.
    This prevents an attack where the same plot ID could be used to create sectors with different
    `history_size` values, which would allow farmers to manipulate shard allocation by assigning to

>     a plot the most convenient sector at each time.

Reviewer: The description is a bit difficult to follow here, I'd simplify it

> 3.  **Load-Balanced Archiving (New History):** As new history arrives, it gets stored across
>     farmers/plots without large re-plotting events.

Reviewer: ```suggestion
3.  **Load-Balanced Archiving (New History):** As new history is archived, it gets plotted shortly after by sufficient number of
    farmers without large re-plotting events.
```

4.  **Detect Expiration and Validity:** Plots, sectors and blocks include all the information needed
    to verify their validity, plot assignment to a shard, sector belonging to the plot, and
    expiration.
5.  **No Excessive Re-plotting:** Keep re-plotting burden similar to original Subspace.

### Committing sectors to history ranges

The core idea for sector expiration is the following: instead of letting farmers commit their
sectors to any history size, they are bound to commit them to a range of history sizes determined by
the plot the sector is linked to. This way, we limit the range of history sizes that a sector can be

> committed to, and hence the parallel sectors with the same `plot_id` that and different committed

Reviewer: "that and" sounds off, probably leftover from a different earlier version

history sizes that can be created. This is implemented as a "self-evolving plot commitment range"
that grows as the blockchain grows on top of the existing mechanism for sector expiration. Thus, we
maintain all the properties of the original Subspace protocol, while adapting it to the needs of the
sharded version. This allows us to also learn an apply the heuristics from the live Subspace
network.

The high-level operation of the protocol is as follows:

> - When creating a new plot, farmers need to choose a random offset (essentially a nonce) for their

Reviewer: I actually like "nonce" as a name for the parameters. Offset explains what it is, but "nonce" seems like a more established terminology for a parameter that user controls in various protocols.

Author: Agree. Renamed throughout the whole document

  plot. This offset determines the specific ID of the plot and will influence the history range that
  the plot's sectors can commit to. As a reminder, plots identities are inferred from the public key
  hash of the farmer, and the random offset they choose:

```rust
plot_id = hash(public_key_hash || offset)
```

- When creating new sectors for a plot, farmers will choose any valid history size within the plot's
  current range. Piece selection is based on the sector's specific history size, and expiration
  follows the same logic as the original Subspace protocol. Finally, the sector ID is derived from
  the plot ID, sector index, and committed history size:

```rust

> sector_id = keyed_hash(plot_id, sector_index || committed_history_size)

Reviewer: They commit to a range, not a specific history size

Author: Does it? I actually was thinking that the sector would commit to a specific history size chosen by the farmer within the range so no changes are required to the expiration logic. We can commit to a history range too, in which case the beginning or the end of the range will determine the history size used for the expiration logic. 

No strong opinions, happy to introduce whatever you think is best (I don't think it changes much)

Reviewer: If they can choose multiple history sizes within a range, they can create multiple valid sectors for the same `sector_index`. The whole point of the history range was to make them commit to a range instead, resulting in a single `sector_id` per sector range instead.

```

- We determine a protocol parameter `BASE_WINDOW_SIZE` that determines the size of the `BASE_RANGE`.
  This `BASE_RANGE` determines the baseline history range for all plots. This `BASE_RANGE`
  determines the size of history sizes that plots can commit to initially. The plot offset modifies
  the `BASE_RANGE` to create a unique range for each plot ID, and as the history of the chain grows,
  this is expanded accordingly so it can fit new history sizes. The following image belongs to a
  [simple visualisation tool](./2025-06-30-sector-expiration/range-evolution-visual.html) that an
  LLM has built for me to visualise how the base range and effective range of a plot behaves as the

>   history of the chain grows:

Reviewer: I know you tried, but I wouldn't be able to understand anything from such description :sweat_smile:

<p align="center">
<img alt="Range evolution visualisation" src="range_visualisation.png">
</p>

- Sectors expire as they are used to in current Subspace protocol, so even if the effective range
  for plots grow, sectors committed to old history sizes will be alreaedy expired (or close to
  expire) still limiting the number of history sizes that sectors can commit to at a given time.
  This is key to ensure that farmers do not need to re-plot their sectors as the history grows, and
  that they can archive the most recent history without having to re-plot, maintaining all the nice
  properties for sector expiration of the original Subspace protocol.

### Putting all pieces together

With this, we have all that we needed to finally implement farming in shards:

- Plot IDs are unique and stable, allowing us to allocate space into shards.
- Sectors are committed to a range of history sizes determined by the plot ID, preventing farmers
  from gaming shard allocation.
- Piece selection, sector expiration stay unchanged.

> - Farming stays slightly unchanged, apart from the additional information required to verify that

Reviewer: ```suggestion
- Farming stays mostly unchanged, apart from the additional information required to verify that
```

  the chunk for the piece in the solution range belongs to a sector bound to a plot currently
  assigned to the shard.

This is how the verification logic for a solution in a shard would look like:

- Standard Subspace Verification (Unchanged): First, perform all existing Subspace checks

```rust
// Verify piece offset is valid
if piece_offset >= max_pieces_in_sector {
    return Err("Invalid piece offset");
}

// Check sector expiration
if sector_is_expired(sector_id, current_history_size) {
    return Err("Sector has expired");
}

// Verify the piece is part of blockchain history
if !verify_piece_inclusion(piece, segment_root, proof) {
    return Err("Invalid piece proof");
}
```

- Plot Identity Verification: Verify that the solution comes from a valid plot:

```rust
// Step 2.1: Compute expected plot ID from farmer's public key and offset
expected_plot_id = hash(farmer_public_key || plot_offset)

// Step 2.2: Verify the sector was created with this plot ID
// The sector_id should incorporate the plot_id
if !sector_belongs_to_plot(sector_id, expected_plot_id) {
    return Err("Sector doesn't belong to claimed plot");

> }

Reviewer: In practice this will be used to derive sector ID and if something doesn't match, it'll blow up somewhere later. We'll not check that sector ID belongs to plot ID as such. In fact there is no sector ID in the solution, it is a derived concept not explicitly present in the solution.

```

- Range Validation: Ensure the sector's committed history size is within the plot's valid range:

```rust
// Step 3.1: Calculate the plot's base range from its offset
window_level = plot_offset / OFFSETS_PER_WINDOW
window_size = BASE_WINDOW_SIZE * (2 ^ window_level)
offset_in_window = plot_offset % OFFSETS_PER_WINDOW

min_history = (window_size * offset_in_window / OFFSETS_PER_WINDOW) - GENESIS_OFFSET

> max_history = (window_size * (offset_in_window + 1) / OFFSETS_PER_WINDOW) - GENESIS_OFFSET

Reviewer: Both of these appear to be fixed for the duration of the plot (assuming `plot_offset` is fixed), I don't think that should be the case.

// Step 3.2: Calculate effective range (extends with blockchain growth)
effective_max_history = max(max_history, current_history_size)

// Step 3.3: Verify sector's history size is within range
if sector_history_size < min_history || sector_history_size > effective_max_history {
    return Err("Sector history size outside plot's valid range");
}
```

- Shard Assignment Verification: Confirm the plot is assigned to the current shard:

```rust
// Step 4.1: Determine which shard this plot belongs to
// Use the plot ID and base range for stable assignment

> expected_shard_id = VRF(plot_id || min_history) % NUM_SHARDS

Reviewer: `plot_id` is a hash, it can't be used to verify VRF. It is more like we'll use VRF signature to determine shard ID and then verify it against public key of the block seal to see if it is valid.

Author: I updated the code, hopefully it is clearer now

// Step 4.2: Verify this matches the shard for the block being verified
if expected_shard_id != current_shard_id {
    return Err("Plot is not assigned to this shard");
}
```

> To perform sharded verification, validators need:

Reviewer: Let's use more generic terminology, this sounds like PoS when "validator" is used:
```suggestion
To perform sharded verification, clients need:
```
